internal/extract.go

```diff
@@ -242,3 +249,27 @@ func tryExtractFiles(files []ReaderMaker,
 	})
 	return failed
 }
+
+func readTrailingZeros(r io.Reader) error {
+	// on first iteration we read small chunk
+	// in most cases we will return fast without memory allocation
+	b := make([]byte, 1024)
+	for {
+		n, err := r.Read(b)
+		if n > 0 {
+			if !utility.AllZero(b[:n]) {
+				return io.ErrClosedPipe
+			}
+		}
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+		if len(b) < utility.CompressedBlockMaxSize {
+			// but if we found zeroes, allocate large buffer to speed up reading
+			b = make([]byte, utility.CompressedBlockMaxSize)
```

Comment:
This condition can fire multiple times, so it's possible to read much more than `utility.CompressedBlockMaxSize`

A `totalBytesRead` counter should be kept, breaking from loop when it reaches `utility.CompressedBlockMaxSize`

Then `utility.CompressedBlockMaxSize - totalBytesRead` would be the allocation size

Reply:
`b` allocated only two times: first with `1024` len, second - with `utility.CompressedBlockMaxSize` len.
Golang zeroes `b` during allocation.  `b`'s len is not affected by `Read` calls.

As we only need to check any non-zero byte, we may omit typical golang Read pattern and always read to the begining of the `b`.

Looks safe to me..

Reply:
Yes, the issue I was considering was the reader giving off more zeroes than `utility.CompressedBlockMaxSize`